clients/apply/apply.go

```diff
@@ -0,0 +1,729 @@
+		if err != nil {
+			return fmt.Errorf("failed to get ID for org %q: %w", orgName, err)
+		}
+		if len(orgs.GetOrgs()) == 0 {
+			return fmt.Errorf("no orgs found with name %q: %w", orgName, err)
+		}
+		orgID = orgs.GetOrgs()[0].GetId()
+	}
+
+	templates, err := readTemplates(ctx, params.Sources)
+	if err != nil {
+		return err
+	}
+
+	req := api.TemplateApply{
+		DryRun:    true,
+		OrgID:     orgID,
+		Templates: templates,
+		EnvRefs:   params.EnvVars,
+		Secrets:   params.Secrets,
+		Actions:   make([]api.TemplateApplyAction, len(params.Filters)),
+	}
+	if params.StackId != "" {
+		req.StackID = &params.StackId
+	}
+	for i, f := range params.Filters {
+		req.Actions[i].Action = api.TEMPLATEAPPLYACTIONKIND_SKIP_KIND
+		if f.Name != nil {
+			req.Actions[i].Action = api.TEMPLATEAPPLYACTIONKIND_SKIP_RESOURCE
+		}
+		req.Actions[i].Properties = api.TemplateApplyActionProperties{
+			Kind:                 f.Kind,
+			ResourceTemplateName: f.Name,
+		}
+	}
+
+	// Initial dry-run to make the server summarize the template & report any missing env/secret values.
+	res, err := c.ApplyTemplate(ctx).TemplateApply(req).Execute()
+	if err != nil {
+		return fmt.Errorf("failed to check template impact: %w", err)
+	}
+
+	if len(res.Summary.MissingEnvRefs) > 0 || len(res.Summary.MissingSecrets) > 0 {
+		for _, e := range res.Summary.MissingEnvRefs {
+			val, err := c.StdIO.GetStringInput(fmt.Sprintf("Please provide environment reference value for key %s", e), "")
+			if err != nil {
+				return err
+			}
+			req.EnvRefs[e] = val
+		}
+		for _, s := range res.Summary.MissingSecrets {
+			val, err := c.StdIO.GetSecret(fmt.Sprintf("Please provide secret value for key %s (optional, press enter to skip)", s), 0)
+			if err != nil {
+				return err
+			}
+			if val != "" {
+				req.Secrets[s] = val
+			}
+		}
+
+		// 2nd dry-run to see the diff after resolving all env/secret values.
+		res, err = c.ApplyTemplate(ctx).TemplateApply(req).Execute()
+		if err != nil {
+			return fmt.Errorf("failed to check template impact: %w", err)
+		}
+	}
+
+	if !params.Quiet {
+		if err := c.printDiff(res.Diff, params); err != nil {
+			return err
+		}
+	}
+
+	if !params.Force {
+		if confirmed := c.StdIO.GetConfirm("Confirm application of the above resources"); !confirmed {
+			return errors.New("aborted application of template")
+		}
+	}
+
+	if !params.OverwriteConflicts && hasConflicts(res.Diff) {
+		return errors.New("template has conflicts with existing resources and cannot safely apply")
+	}
+
+	// Flip the dry-run flag and apply the template.
+	req.DryRun = false
+	res, err = c.ApplyTemplate(ctx).TemplateApply(req).Execute()
+	if err != nil {
+		return fmt.Errorf("failed to apply template: %w", err)
+	}
+	params.StackId = res.StackID
+
+	if !params.Quiet {
+		if err := c.printSummary(res.Summary, params); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func readTemplates(ctx context.Context, sources []TemplateSource) ([]api.TemplateApplyTemplate, error) {
```

Comment:
I expect that some of these helper functions will move to separate files so they can be reused in the remaining template-related commands.